### PR TITLE
Do not keep map packages loaded on Game start to reduce memory impact

### PR DESCRIPTION
**Pull request by cclecle**
Address the following issue:
#21407

### CODE REVIEW

File: OpenRA.Game/Map/MapCache.cs [hunk 1]
```diff
@@ -128,7 +128,10 @@ public void LoadMaps()
 			foreach (var kv in MapLocations)
 			{
 				foreach (var map in kv.Key.Contents)
+				{
 					LoadMapInternal(map, kv.Key, kv.Value, mapGrid, null, modDataRules);
+					GC.Collect();
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapCache.cs]: this is a foreach loop, perhaps it should be moved to be after foreach instead for better perf?

File: OpenRA.Game/Map/MapCache.cs [hunk 2]
```diff
@@ -148,10 +151,14 @@ void LoadMapInternal(string map, IReadOnlyPackage package, MapClassification cla
 				using (new PerfTimer(map))
 				{
 					mapPackage = package.OpenPackage(map, modData.ModFiles);
+					Log.Write("debug",$"Loading Map: {map} FROM {mapPackage.Name}");
+					Console.WriteLine($"Loading Map: {map} FROM {mapPackage.Name}");
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapCache.cs]: please don't include debugging in the PR

File: OpenRA.Game/Map/MapCache.cs [hunk 3]
```diff
@@ -168,9 +175,13 @@ void LoadMapInternal(string map, IReadOnlyPackage package, MapClassification cla
 				Console.WriteLine($"Failed to load map: {map}");
 				Console.WriteLine("Details:");
 				Console.WriteLine(e);
+				Console.WriteLine("StackTrace:");
+				Console.WriteLine(System.Environment.StackTrace);
 				Log.Write("debug", $"Failed to load map: {map}");
 				Log.Write("debug", "Details:");
 				Log.Write("debug", e);
+				Log.Write("debug", "StackTrace:");
+				Log.Write("debug",System.Environment.StackTrace);
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapCache.cs]: not including the tracelog is intentional, we do not want to spam users with useless data. We provide helpful errors via linting.

File: OpenRA.Game/Map/MapPreview.cs [hunk 4]
```diff
@@ -174,7 +174,20 @@ public InnerData Clone()
...
+				if (_Package == null)
+				{
+					_Package = parentPackage.OpenPackage(PackageName, modData.ModFiles);
+				}
+				return _Package;
+			}
+			private set {
+				_Package = value;
+			}
+		}
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapPreview.cs:188]: This structure doesn't seem fully correct. MapDirectoryTracker cycles through all packages so it will re-create them. There are also a few places that want just the name of the package but now will open the package

File: OpenRA.Game/Map/MapCache.cs (hunk 1 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapCache.cs]: I guess its a compromise between CPU or memory overhead.
But to be honest I put that as a first attempt to reduce memory usage, it might not be necessary anymore.
I will try without and if it work I will just drop it. I dont think it make sense to put it outside the loop because the memory big-bang happened within that loop.

File: OpenRA.Game/Map/MapCache.cs (hunk 2 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapCache.cs]: right, I'll drop these lines.I just thought it could be usefull for users.

File: OpenRA.Game/Map/MapCache.cs (hunk 3 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapCache.cs]: Got it.

File: OpenRA.Game/Map/MapPreview.cs (hunk 4 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs:188]: This is the "core" of the change. 
Packages are only loaded on-demand and then stay loaded if not explicitly unloaded using PackageDispose().
So in the first loadmaps loop I explicitly unload them by calling the PackageDispose function, but then if later the app code try to access them again then yes they will be reloaded. But I guess that's exactly what we want ?

Comment (PunkPun) [reply OpenRA.Game/Map/MapPreview.cs:188]: the issue is that as soon the the tracker detects a change the packages will be re-created. And also packages are created when there isn't a need

Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs:188]: I see, but they were already created when there isn't a need. I just made them unloaded after first mapload.
So as I see the tracker is using Package.Name, wich will reload the package. 
I can use the new  PackageName field from MapPreview instead of the one in the Package to prevent it from loading the Package again ?

File: OpenRA.Game/Map/MapCache.cs (hunk 1 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapCache.cs]: removed

File: OpenRA.Game/Map/MapPreview.cs (hunk 4 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs:188]: I searched for .Package in the whole source code and fix this one and another one. Should be fine ? An alternative would be to create a getPackage() method that gives the loaded package without keeping the reference but that would break the MapPreview class behavior because it is using Package internally to implement IReadOnlyFileSystem interface. 
So current change proposal looks like a good compromise ?

File: OpenRA.Game/Map/MapPreview.cs [hunk 5]
```diff
@@ -284,9 +298,10 @@ public MapPreview(Map map, ModData modData)
 		{
 			this.modData = modData;
 			cache = modData.MapCache;
+			this.pPackage = null;
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapPreview.cs]: this is null by default, no need to set it

File: OpenRA.Game/Map/MapPreview.cs [hunk 6]
```diff
@@ -258,6 +270,8 @@ public MapPreview(ModData modData, string uid, MapGridType gridType, MapCache ca
 		{
 			this.cache = cache;
 			this.modData = modData;
+			this.pPackage = null;
+			this.PackageName = null;
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapPreview.cs]: same here

File: OpenRA.Game/Map/MapPreview.cs (hunk 5 above)
Comment (PunkPun) [reply OpenRA.Game/Map/MapPreview.cs]: also rename to just `package`, there's no need for a special name

File: OpenRA.Game/Map/MapPreview.cs (hunk 4 above)
Comment (PunkPun) [reply OpenRA.Game/Map/MapPreview.cs:188]: Screenshot 2024-04-24 at 17 16 35

there's still a few more. One in MissionBrowser, it would load all packages on entering the menu. Another in SaveMapLogic, it would do the same.

File: OpenRA.Game/Map/MapCache.cs [hunk 7]
```diff
@@ -167,7 +172,7 @@ void LoadMapInternal(string map, IReadOnlyPackage package, MapClassification cla
 				mapPackage?.Dispose();
 				Console.WriteLine($"Failed to load map: {map}");
 				Console.WriteLine("Details:");
-				Console.WriteLine(e);
+				Console.WriteLine(System.Environment.StackTrace);
```
Comment (PunkPun) [inline OpenRA.Game/Map/MapCache.cs]: unnecessary change

File: OpenRA.Game/Map/MapPreview.cs (hunk 5 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs]: ok

File: OpenRA.Game/Map/MapPreview.cs (hunk 6 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs]: ok

File: OpenRA.Game/Map/MapCache.cs (hunk 7 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapCache.cs]: woups, that's a mistake, ok

File: OpenRA.Game/Map/MapPreview.cs (hunk 4 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs:188]: done

File: OpenRA.Game/Map/MapPreview.cs (hunk 5 above)
Comment (cclecle) [reply OpenRA.Game/Map/MapPreview.cs]: done

File: OpenRA.Game/Map/MapCache.cs [hunk 8]
```diff
@@ -128,7 +128,9 @@ public void LoadMaps()
 			foreach (var kv in MapLocations)
 			{
 				foreach (var map in kv.Key.Contents)
+				{
 					LoadMapInternal(map, kv.Key, kv.Value, mapGrid, null, modDataRules);
+				}
```
Comment (abcdefg30) [inline OpenRA.Game/Map/MapCache.cs]:
```suggestion
					LoadMapInternal(map, kv.Key, kv.Value, mapGrid, null, modDataRules);
```

File: OpenRA.Game/Map/MapPreview.cs [hunk 9]
```diff
@@ -587,10 +599,15 @@ public void Invalidate()
 
 		public void Dispose()
 		{
-			if (Package != null)
+			PackageDispose();
+		}
+
+		public void PackageDispose()
```
Comment (abcdefg30) [inline OpenRA.Game/Map/MapPreview.cs]:
```suggestion
		public void DisposePackage()
```

**Review by RoosterDragon — APPROVED** [review]

**Review by PunkPun — APPROVED** [review]
LGTM

**Comment by PunkPun** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ccc25513655cdf98989aa2ff10554c830eee3957...42ab9db2f62aa3427e66c1042d8df815235cac93)

**Comment by PunkPun** [comment]
I think it would make sense to add package disposal on game end as to not be bloating the ram over time. Or on server start we could dispose of all but the selected map's package

**Comment by cclecle** [comment]
thanks for the merge :)
I will have a look for the last comment.
